src/Workspaces/CSharp/Portable/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs

```diff
@@ -1116,25 +1116,46 @@ private IEnumerable<ITypeSymbol> InferTypeInIfStatement(IfStatementSyntax ifStat
                 ExpressionSyntax expressionOpt = null,
                 SyntaxToken? previousToken = null)
             {
-                if (initializerExpression.IsParentKind(SyntaxKind.ImplicitArrayCreationExpression))
+                if (initializerExpression.IsKind(SyntaxKind.ComplexElementInitializerExpression))
                 {
-                    // First, try to infer the type that the array should be.  If we can infer an
-                    // appropriate array type, then use the element type of the array.  Otherwise,
-                    // look at the siblings of this expression and use their type instead.
+                    // new Dictionary<K,V> { { x, ... } }
+                    // new C { Prop = { { x, ... } } }
+                    var parameterIndex = previousToken.HasValue
+                            ? initializerExpression.Expressions.GetSeparators().ToList().IndexOf(previousToken.Value) + 1
+                            : initializerExpression.Expressions.IndexOf(expressionOpt);
 
-                    var arrayTypes = this.InferTypes((ExpressionSyntax)initializerExpression.Parent);
-                    var elementTypes = arrayTypes.OfType<IArrayTypeSymbol>().Select(a => a.ElementType).Where(e => !IsUnusableType(e));
+                    var addMethodSymbols = _semanticModel.GetCollectionInitializerSymbolInfo(initializerExpression).GetAllSymbols();
+                    var addMethodParameterTypes = addMethodSymbols
+                        .Where(a => ((IMethodSymbol)a).Parameters.Count() == initializerExpression.Expressions.Count)
+                        .Select(a => ((IMethodSymbol)a).Parameters.ElementAtOrDefault(parameterIndex)?.Type)
```

Comment:
To check: ElementAtOrDefault is doing a fast lookup rather than O(n)? Ditto for the Count() above?